### PR TITLE
Move `lint_store`

### PR DESCRIPTION
**Pull request by nnethercote**
Some nice cleanups enabled by the removal of compiler plugins.

r? @cjgillot

### CODE REVIEW

File: compiler/rustc_interface/src/interface.rs
Comment (tamird) [inline compiler/rustc_interface/src/interface.rs:453]: I think this is now the only caller of `Session::enable_internal_lints`, which means its logic can be inlined above, and then the lint store created at the same time as the session. That would enable the lint store to be non-optional.

Comment (nnethercote) [reply compiler/rustc_interface/src/interface.rs:453]: The lint store could be created, but the initialization via `register_lints` would have to wait until after the `Session` is built. And by that stage the lint store is held in `Session` as `Lrc<dyn Any + DynSync + DynSend>`, which is hard to modify and is intended to be a read-only form.

Comment (tamird) [reply compiler/rustc_interface/src/interface.rs:453]: Is there a reason it has to be held as `Lrc<dyn Any + DynSync + DynSend>` rather than by its concrete type (with or without `Lrc`)?

Comment (nnethercote) [reply compiler/rustc_interface/src/interface.rs:453]: Yes. `rustc_session` does not (and cannot) depend on `rustc_lint`.

Comment (nnethercote) [reply compiler/rustc_interface/src/interface.rs:453]: As the comment on `lint_store` says:
```
/// This only ever stores a `LintStore` but we don't want a dependency on that type here.
```

**Comment by nnethercote** [comment]
> The job **`x86_64-gnu-tools`** failed!

Two weird clippy failures.
- In `src/tools/clippy/tests/ui/macro_use_imports.rs`, the four errors are being emitted in a different order.
- In `src/tools/clippy/tests/ui/dbg_macro.rs` we're getting a slightly different suggestion in one place, due to somehow hitting the `else` branch instead of the `if` branch in [this code](https://github.com/rust-lang/rust/blob/7b97a5ca8422d1495a8918106d3249aa405812d4/src/tools/clippy/clippy_lints/src/dbg_macro.rs#L91-L97).

I bisected locally and found it's third commit that's causing these changes, but I don't know why.

**Comment by Alexendoo** [comment]
It's likely because spans are changing, clippy loads `clippy.toml` into the source map in the `register_lints` callback

`dbg_macro` handled them improperly and will be fixed by https://github.com/rust-lang/rust-clippy/pull/11743

`macro_use_imports` keys on a span among other things in an `FxHashMap`, haven't opened a PR for this one yet

Blessing them would be fine, we can fix the other one on the clippy side

**Comment by nnethercote** [comment]
> It's likely because spans are changing, clippy loads `clippy.toml` into the source map in the `register_lints` callback
> 
> `dbg_macro` handled them improperly and will be fixed by [rust-lang/rust-clippy#11743](https://github.com/rust-lang/rust-clippy/pull/11743)
> 
> `macro_use_imports` keys on a span among other things in an `FxHashMap`, haven't opened a PR for this one yet
> 
> Blessing them would be fine, we can fix the other one on the clippy side

Thanks for the info, that's a huge help!

I had suspected the `macro_use_imports` one was an `FxHashMap` ordering thing. Good to have that confirmed.

The `dbg_macro` one was indeed a span issue, I found that the offset for a particular span had changed by about 70 bytes. I didn't think about the offset operation being bogus.

If I remove `src/tools/clippy/clippy.toml` and `src/tools/clippy/tests/clippy.toml` then all the clippy tests pass.

But I don't understand how my change affected clippy's use of `register_lints`, which concerns me. This change wasn't supposed to have any visible effects.

**Comment by Alexendoo** [comment]
Haven't finished it yet but a local clippy change I have to read the config file in [`Config::hash_untracked_state`](https://github.com/rust-lang/rust/pull/116731) broke these same two tests, I believe the cause would be the same here

Because `register_lints` is being called earlier it's loading `clippy.toml` into the source map before e.g. `dbg_macro.rs`, meaning `BytePos` 0 no longer points to the start of the test file

**Comment by nnethercote** [comment]
Thanks, @Alexendoo. I have updated the expected test outputs. This should now be ready to review.

**Comment by blyxyas** [comment]
This PR would totally block and mean a total re-write of the current efforts for not processing  allowed lints (both in Clippy and Rustc) as it removed practically everything that the current version of that patch uses.

I advise against this cleanup, at least until that patch is open and merged.

File: src/tools/clippy/tests/ui/dbg_macro.stderr
Comment (flip1995) [inline src/tools/clippy/tests/ui/dbg_macro.stderr:111]: The Clippy stderr file changes look really sus. Why did they change by just moving the `lint_store` to `Session`?

**Review by flip1995 — CHANGES_REQUESTED** [review]

**Comment by nnethercote** [comment]
> This PR would totally block and mean a total re-write of the current efforts for not processing allowed lints (both in Clippy and Rustc) as it removed practically everything that the current version of that patch uses.

Do you have a link to that patch?

File: src/tools/clippy/tests/ui/dbg_macro.stderr
Comment (flip1995) [reply src/tools/clippy/tests/ui/dbg_macro.stderr:111]: Ah, just saw the conversation above. I'd rather not merge this Clippy regression, until the fixes for this in Clippy are synced to the Rust repo.

Comment (nnethercote) [reply src/tools/clippy/tests/ui/dbg_macro.stderr:111]: As @Alexandoo described [above](https://github.com/rust-lang/rust/pull/117649#issuecomment-1798543616), that test is badly written and this change just exposes the latent badness.

**Comment by nnethercote** [comment]
> > This PR would totally block and mean a total re-write of the current efforts for not processing allowed lints (both in Clippy and Rustc) as it removed practically everything that the current version of that patch uses.
> 
> Do you have a link to that patch?

@flip1995 pointed me at #106983, which has an attempted fix in #116271. I looked through the code for #116271 and I don't see any real overlap with the changes in this PR. So I think "totally block and mean a total re-write" is untrue.

File: src/tools/clippy/tests/ui/dbg_macro.stderr
Comment (nnethercote) [reply src/tools/clippy/tests/ui/dbg_macro.stderr:111]: What's the mechanism and timing for Clippy syncing into the Rust repo?

**Comment by blyxyas** [comment]
I wasn't talking about #116271, I was talking about a more modern form of [this patch](https://github.com/rust-lang/rust/compare/master...blyxyas:rust:ignore-allowed-lints) (which I lost and will have to rewrite from memory).

And blocking that patch would block the Clippy efforts to ignore allowed lints (apart from Alexendoo's patch, which I don't know the progress of, or usage of this objects)

Some things can indeed be cleaned up, but others like `LintContext` or `get_lints()` being derived on each lint cannot.

**Comment by nnethercote** [comment]
> I wasn't talking about #116271

Your original comment said this:

> This PR would totally block and mean a total re-write of the current efforts for not processing allowed lints (both in Clippy > and Rustc) as it removed practically everything that the current version of that patch uses.

Strong words! But there was no link to the patch and you didn't respond to my request for a link to it. So I contacted @flip1995 on Zulip and they pointed me at #116271.

> I was talking about a more modern form of [this patch](https://github.com/rust-lang/rust/compare/master...blyxyas:rust:ignore-allowed-lints) (which I lost and will have to rewrite from memory).

Ok, that's helpful. I just looked at that patch and... again, it doesn't seem to conflict with this PR at all?

> And blocking that patch would block the Clippy efforts to ignore allowed lints (apart from Alexendoo's patch, which I don't know the progress of, or usage of this objects)
> 
> Some things can indeed be cleaned up, but others like `LintContext` or `get_lints()` being derived on each lint cannot.

This PR removes `LintContext::PassObject` and `LintContext::lints`. It doesn't touch `get_lints`.

I don't want to get in people's way and I'm happy to make adjustments if there are specific changes here that cause problems. But I don't see any such problems so far. I can't read minds so I don't know what was in past patches that have been lost, nor what will be in future patches yet to be written. I feel like we can all work together to get most or all of this PR merged.

**Comment by blyxyas** [comment]
On an even closer inspection, the current patch indeed isn't blocked by this PR, let's hope that the next iteration isn't as well.
Sorry for the false alarm.

File: compiler/rustc_interface/src/interface.rs
Comment (cjgillot) [inline compiler/rustc_interface/src/interface.rs:455]: There are a few initialization routines that change sess after this call. What do callbacks expect to see? Should this initialization happen a bit later in this function?

**Review by cjgillot — COMMENTED** [review]
Just one question. r=me if no issue.

File: compiler/rustc_interface/src/interface.rs
Comment (nnethercote) [reply compiler/rustc_interface/src/interface.rs]: Good suggestion. I've moved the initialization to just before the session is wrapped in an `Lrc` and moved into the `Compiler`.

**Comment by nnethercote** [comment]
I've addressed @cjgillot's comment. I will wait until the next clippy sync (Thursday) before merging this, to avoid the problem with `dbg_macro.rs`'s expected output.

File: src/tools/clippy/tests/ui/dbg_macro.stderr
Comment (nnethercote) [reply src/tools/clippy/tests/ui/dbg_macro.stderr]: The next sync is on Thursday.

**Comment by nnethercote** [comment]
The clippy sync has occurred, and I have updated.

@bors r=cjgillot

**Pull request merged**
